### PR TITLE
feat(prenotes): Add the ability to pass prenote messages between plugins

### PR DESCRIPTION
**Pull request by AndyTWF**

### CODE REVIEW

File: database/migrations/2021_07_04_153303_add_prenotes_columns_to_controller_positions_table.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            $table->boolean('receives_prenotes')
+                ->after('sends_prenotes')
+                ->default(false)
+                ->comment('Whether the controller position can receive prenotes');
+        });
+
+        DB::table('controller_positions')
+            ->update(
+                [
+                    'sends_prenotes' => DB::raw('`requests_departure_releases`'),
```
Comment (AxonC) [inline database/migrations/2021_07_04_153303_add_prenotes_columns_to_controller_positions_table.php:32]: What is this needed for?

File: routes/api.php [hunk 2]
```diff
@@ -73,6 +74,15 @@ function () {
                 Route::put('notifications/read/{id}', 'NotificationController@readNotification')
                     ->where('id', '[0-9]+');
 
+                // Prenotes
+                Route::prefix('prenote-messages')->group(function () {
```
Comment (AxonC) [inline routes/api.php:78]: If this was entirely RESTful, shouldn't this be `prenotes/messages`?

**Review by AxonC — CHANGES_REQUESTED** [review]

File: database/factories/Airfield/UsesAirfieldIcaoCodes.php [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace Database\Factories\Airfield;
+
+use Faker\Generator;
+use Illuminate\Support\Str;
+
+trait UsesAirfieldIcaoCodes
+{
+    private function getAirfieldIcao(Generator $faker): string
```
[the remark this answers is not in the document]
Comment (AndyTWF) [reply database/factories/Airfield/UsesAirfieldIcaoCodes.php:10]: I liked this!

File: database/migrations/2021_07_04_153303_add_prenotes_columns_to_controller_positions_table.php (hunk 1 above)
Comment (AndyTWF) [reply database/migrations/2021_07_04_153303_add_prenotes_columns_to_controller_positions_table.php:32]: So the database fields the controller-positions dependency in the plugin. It's essentially to improve the UX plugin-side. It prevents positions that can't receive prenotes from appearing in the controller list / prevents people who can't request from requesting.

In theory, if someone can approve a release, they can probably accept a prenote, so this is a good starting point :)

File: routes/api.php (hunk 2 above)
Comment (AndyTWF) [reply routes/api.php:78]: Good shout, 8dcf7cfc77abc6ae89cf0647bd75e91715eeda27

File: database/factories/Prenote/PrenoteMessageFactory.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+     *
+     * @return array
+     */
+    public function definition()
+    {
+        return [
+            'callsign' => 'BAW123',
+            'departure_airfield' => $this->getAirfieldIcao($this->faker),
+            'user_id' => $this->activeUser()->id,
+            'controller_position_id' => ControllerPosition::factory()->create()->id,
```
[the remark this answers is not in the document]
Comment (AndyTWF) [reply database/factories/Prenote/PrenoteMessageFactory.php:34]: It actually does! ba185e1b

File: app/Http/Controllers/PrenoteMessageController.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                Auth::id(),
+                $request->validated()['controller_position_id']
+            );
+        } catch (PrenoteAcknowledgementNotAllowedException $notAllowedException) {
+            return response()->json(['message' => $notAllowedException->getMessage()], 403);
+        } catch (PrenoteAlreadyAcknowledgedException $alreadyAcknowledgedException) {
+            return response()->json(['message' => $alreadyAcknowledgedException->getMessage()], 409);
+        }
+
+        return response()->json();
```
[the remark this answers is not in the document]
Comment (AndyTWF) [reply app/Http/Controllers/PrenoteMessageController.php:56]: df39509e

File: app/Services/PrenoteMessageService.php [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+{
+    public function createPrenoteMessage(
+        string $callsign,
+        string $departureAirfield,
+        ?string $departureSid,
+        ?string $destinationAirfield,
+        int $userId,
+        int $sendingControllerId,
+        int $targetControllerId,
+        int $expiresInSeconds
```
[the remark this answers is not in the document]
Comment (AndyTWF) [reply app/Services/PrenoteMessageService.php:24]: Made it slightly less painful! 2a0c7cdf

**Review by AxonC — APPROVED** [review]

**Pull request merged**